src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -167,66 +199,99 @@ def non_pants_project_abs_path(self, path: str) -> Path | None:
             )
 
 
+def produce_match(sep: str, feat: Any) -> str:
+    return f"{sep}{feat}" if feat else ""
+
+
+def add_markers(base: str, attributes: PyprojectAttr, fp) -> str:
+    markers_lookup = produce_match("", attributes.get("markers"))
+    python_lookup = parse_python_constraint(attributes.get("python"), fp)
+
+    # Python constraints are passed as a `python_version` environment marker; if we have multiple
+    # markers, we evaluate them as one whole, and then AND with the new marker for the Python constraint.
+    # E.g. (marker1 AND marker2 OR marker3...) AND (python_version)
+    # rather than (marker1 AND marker2 OR marker3 AND python_version)
+    return (
+        f"{base}"
+        f"{';' if len(markers_lookup) > 0 else ''}"
+        f"{'(' if (' and ' in markers_lookup or ' or ' in markers_lookup) else ''}"
+        f"{markers_lookup}"
+        f"{';' if python_lookup and not markers_lookup else ''}"
+        f"{')' if (' and ' in markers_lookup or ' or ' in markers_lookup) else ''}"
+        f"{' and ' if python_lookup and markers_lookup else ''}"
+        f"{python_lookup}"
```

Comment:
This is a little hard to follow. I think it could benefit from early returns and dedicated conditional branches. For example:

```python
if not markers_lookup and not python_lookup:
   return base
result = base
if markers_lookup:
    result += "; ..."
if python_lookup:
   result += "; ..."
return result